src/directives/offense/poisonRoom.ts

```diff
@@ -1,169 +1,188 @@
-// 			}
-//
-// 			//remove any hostile consituction sites
-// 			_.forEach(this.room.find(FIND_HOSTILE_CONSTRUCTION_SITES), csite => {csite.remove();});
-//
-// 			if (this.isPoisoned()) {
-// 				//remove roads before unclaiming
-// 				if(this.room.roads.length > 0){
-// 					_.forEach(this.room.roads, road => {road.destroy();} );
-// 				}
-// 				this.room.controller!.unclaim();
-// 				//log.notify(`Removing poisonRoom directive in ${this.pos.roomName}: operation completed.`);
-// 				//this.remove();
-// 			} else {
-// 				this.poison();
-// 			}
-// 		}
-//
-// 		// Remove if owned by other player
-// 		if (Game.time % 10 == 2 && this.room && !!this.room.owner && this.room.owner != MY_USERNAME) {
-// 			log.notify(`Removing poisonRoom directive in ${this.pos.roomName}: room already owned by another player.`);
-// 			this.remove();
-// 		}
-// 	}
-// }
-//
-//
-//
-//
-//
+import {Colony, ColonyMemory, getAllColonies} from '../../Colony';
+import {log} from '../../console/log';
+import {ClaimingOverlord} from '../../overlords/colonization/claimer';
+import {RoomPoisonerOverlord} from '../../overlords/offense/roomPoisoner';
+import {StationaryScoutOverlord} from '../../overlords/scouting/stationary';
+import {profile} from '../../profiler/decorator';
+import {Cartographer, ROOMTYPE_CONTROLLER} from '../../utilities/Cartographer';
+import {printRoomName} from '../../utilities/utils';
+import {MY_USERNAME} from '../../~settings';
+import {DirectiveOutpostDefense} from '../defense/outpostDefense';
+import {Directive} from '../Directive';
+import {DirectiveControllerAttack} from './controllerAttack';
+
+export const RUN_TIMER = 25;
+
+/*
+* Poison sources in remote rooms by claiming controller, walling in its sources and controller then unclaiming it.
+*/
+@profile
+export class DirectivePoisonRoom extends Directive {
+	static directiveName = 'PoisonRoom';
+	static color = COLOR_RED;
+	static secondaryColor = COLOR_BROWN;
+	static requiredRCL = 4;
+	walkableSourcePosisions: RoomPosition[];
+	walkableControllerPosisions: RoomPosition[];
+	overlords: {
+		claim: ClaimingOverlord;
+		roomPoisoner: RoomPoisonerOverlord;
+	};
+	constructor(flag: Flag) {
+		super(flag, colony => colony.level >= DirectivePoisonRoom.requiredRCL);
+		// Remove if misplaced
+		if (Cartographer.roomType(this.pos.roomName) != ROOMTYPE_CONTROLLER) {
+			log.warning(`${this.print}: ${printRoomName(this.pos.roomName)} is not a controller room; ` +
+						`removing directive!`);
+			this.remove(true);
+		}
+		// failsafe - if already owned and controller level > 2, remove flag
+		if(this.room && this.room.controller && this.room.controller.my && this.room.controller.level > 2) {
+			this.remove(true);
+		}
+		// Remove if owned by other player
+		if (this.room && !!this.room.owner && this.room.owner != MY_USERNAME) {
+			log.notify(`Removing poisonRoom directive in ${this.pos.roomName}: room already owned by another player.`);
+			this.remove(true);
+		}
```

Comment:
What happens if you don't have vision of a room? (this.room == null)
Do these failsafes work?

Reply:
Auto placing of flag can only be done in rooms with vision, so this should work.
Now if it is manual placements by a player then:
1) in own room = I already have vision
2) in enemy owned room = there is no harm so once a creep arrives, the flag will be removed.

May be u can suggest a better place to have these checks in. currently they are when the flag is placed and everytime the code is reset by screeps servers (around 10 ticks or so i guess).

regarding the last check for GCL, may be I should remove this one because once the room is claimed, the check will fail when GCL is reached.

Reply:
> Auto placing of flag can only be done in rooms with vision, so this should work

Directive.create(pos) can be done in rooms without vision. If the room is not visible, the flag memory gains a `setPosition` property that acts as its effective room position and room and it is moved once vision is gained.